### PR TITLE
[p4fmt]: Add a pretty printer for p4fmt. 

### PR DESCRIPTION
**Pull request by snapdgn**
pretty printer for p4fmt

### CODE REVIEW

**Review by fruffy — APPROVED** [review]

**Review by fruffy — COMMENTED** [review]
I would call it a different name. P4Formatter? Also feel free to prune as many members and methods as possible. I think most of them are not needed here.

Also add a preorder function for `const IR::Node *` throw a P4C_UNIMPLEMENTED exception there.

**Comment by snapdgn** [comment]
> Also add a preorder function for `const IR::Node *` throw a P4C_UNIMPLEMENTED exception there.

Yes, was planning to do this.

File: backends/p4fmt/toP4/toP4.h
Comment (qobilidop) [inline backends/p4fmt/toP4/toP4.h:10]: I feel it's better not to introduce new nested namespaces unless absolutely necessary. Also there might be ambiguity between `::P4Fmt::P4` and `::P4` in unintentional ways. For now, put all P4 formatter code under `::P4Fmt` should be good enough.

**Comment by snapdgn** [comment]
> I would call it a different name. P4Formatter?

Just to clarify things, did you mean I should rename 'toP4' class to 'P4Formatter' ?

**Comment by fruffy** [comment]
> > I would call it a different name. P4Formatter?
> 
> Just to clarify things, did you mean I should rename 'toP4' class to 'P4Formatter' ?

Yes, or whatever name you find most appropriate.

**Review by fruffy — APPROVED** [review]

**Comment by fruffy** [comment]
Need to update the namespace after #4825 has been merged, then we are good to go.

**Comment by fruffy** [comment]
@snapdgn is this ready to be merged?

**Comment by snapdgn** [comment]
> @snapdgn is this ready to be merged? It is still in draft mode.

Sorry, this should be ready now.

File: backends/p4fmt/p4formatter/p4formatter.h
Comment (qobilidop) [inline backends/p4fmt/p4formatter/p4formatter.h:1]: Is this nested directory `p4formatter/p4formatter.h` necessary? How about simply `p4formatter.h`?

**Review by qobilidop — APPROVED** [review]
LGTM other than the directory structure.

File: backends/p4fmt/p4formatter/p4formatter.h
Comment (snapdgn) [reply backends/p4fmt/p4formatter/p4formatter.h:1]: I agree. I'll do away with the nested dir.

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:206]: This looks like a common pattern (visit some collection with comma separator and this `fist` /  not-`first`). Consider refactoring into some common code of visitation with separator. E.g. you can submit a lambda to visit the internals, e.g.:
```c++
visit_collection(t->components, ",", 
                 [](const auto *elem) {
                      auto p4type = a->getP4Type();
                      CHECK_NULL(p4type);
                      visit(p4type);
                   });
```

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:270]: See above

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:285]: ditto

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:417]: `unordered_map`

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:598]: ditto

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:752]: can't you use structured bindings to simplify this?

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:784]: see above

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:813]: ditto

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:835]: ditto

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:903]: again great example of possible RAII usage

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1087]: see above

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1186]: ditto

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1213]: ditto

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1231]: ditto

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1235]: do not use top-level qualified P4 namespace

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1304]: see above

File: backends/p4fmt/p4formatter.cpp
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1402]: `unordered_map`

**Review by asl — DISMISSED** [review]
See comments

File: backends/p4fmt/p4formatter.h
Comment (fruffy) [inline backends/p4fmt/p4formatter.h:10]: You do not need this, just use the fully qualified path.

File: backends/p4fmt/p4formatter.h
Comment (fruffy) [inline backends/p4fmt/p4formatter.h:107]: Nit: `//` instead of `/*`

File: backends/p4fmt/p4formatter.h
Comment (fruffy) [inline backends/p4fmt/p4formatter.h:149]: This is inherited from ToP4 but should not be needed.

**Review by fruffy — APPROVED** [review]
@asl I will go ahead and merge this today if you do not have any other comments.

**Review by qobilidop — APPROVED** [review]

**Comment by fruffy** [comment]
Dismissing @asl's review to merge, we can address remaining issues in subsequent PRs.

**Comment by snapdgn** [comment]
> Most comments have been addressed.

Yes, I don't think I missed resolving any of the suggestions.

**Pull request merged**
